research/src/pages/popup/popup.proposal.mdx

```diff
@@ -0,0 +1,342 @@
+  purposes of deciding whether to hide a popup.
+</p>
+
+<p class="question">
+  Should we have a way to suppress these steps upon focus, so that the author can enable the popup
+  to be shown upon text entry (or some other logic) instead?
+</p>
+
+<h3 id="showing-popup-steps">
+  Showing a <code>popup</code> element steps
+</h3>
+
+1. Run the [hiding currently-shown `popup` elements steps](#hiding-current-popups).
+2. Add the _subject_ to _subject_'s
+   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s
+   [top layer](https://fullscreen.spec.whatwg.org/#top-layer), at the top of the stack.
+3. Set the `open` IDL attribute on the _subject_ to `true`.
+
+<h4 id="hiding-current-popups">
+  Hiding currently-shown <code>popup</code> elements steps
+</h4>
+
+Let _candidate subject_ be the new candidate `popup` element to be shown.
+If applicable, let _activating element_ be the element whose `popup` attribute specifies the ID of
+the _candidate subject_, and which has been activated by the user.
+If applicable, let the _anchor element_ be the element whose ID is referenced by the `anchor`
+specified on the _candidate subject_.
+
+Starting at the bottom of the top-layer stack: for each `popup` element which has an `open` IDL
+attribute set to `true`, and which is not the _candidate subject_:
+
+1. Let _subject_ be the current `popup` element in the top-layer stack.
+2. Walk the DOM to determine if the _subject_ is an ancestor to the _candidate subject_, the
+   _activating element_, or the _anchor element_.
+3. If the _subject_ is an ancestor, then return.
+4. Otherwise, for this _subject_ and any other `popup`s higher in the stack, run the
+   [hiding a `popup` element steps](#hiding-current-popups).
+
+<p class="todo">
+  Step #2 probably needs to be more precise, and more specific about Shadow/light DOM
+</p>
+
+## Setting focus when the `popup` element is shown
+
+### The `delegatesfocus` attribute
+
+The `delegatesfocus` attribute is a
+[boolean attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute).
+When specified, it indicates that when the `popup` element is shown, focus should move to the first
+focusable descendent of the `popup` element.
+
+**Example:**
+
+```html
+<popup delegatesfocus>
+  <p>I am not a focusable element.</p>
+  <p>Nor am I.</p>
+  <button>I will be focused whenever the popup becomes focused.</button>
+</popup>
+```
+
+### Focusing steps
+
+The popup [focusing steps](https://html.spec.whatwg.org/multipage/interaction.html#focusing-steps)
+for a `popup` element _subject_ are as follows:
+
+1. If the `delegatesfocus` attribute is specified on _subject_, let _control_ be the first focusable
+   descendant element of _subject_, in [tree order](https://dom.spec.whatwg.org/#concept-tree-order),
+   that is not inert and has the
+   [autofocus](https://html.spec.whatwg.org/multipage/interaction.html#attr-fe-autofocus) attribute
```

Comment:
This is a bit different than I [implemented](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_popup_element.cc;l=155;drc=1de0bd095828489ebfef0c265e172af3d1dd0afd) in Chromium, in that the `delegatesfocus` attribute causes the first focusable element to be focused, regardless of whether it **also** has the `autofocus` attribute. If `subject` does **not** have the `delegatesfocus` attribute, **then** the code looks for a descendant element that has the `autofocus` attribute, and focuses that.

I can change that behavior, but it actually makes a bit more sense to me than the algorithm here, as written. As written, you need to **both** add `delegatesfocus` **and** add `autofocus` to an element, which means there's no way to just say "focus the first thing you can within this popup". It feels like these should be two separate features:
 1. `delegatesfocus` on the `<popup>` itself: focus the first focusable thing inside
 2. `autofocus` on the `<popup>` or any contained element: focus that

Reply:
Honestly, I think I just missed moving some text around, because you're right, the algo as written makes no sense. Sorry about that and thank you for flagging it! Should be fixed in this commit: https://github.com/openui/open-ui/pull/355/commits/252119e3d1274313d70f6bf4fb5b21ecfb0d587b